src/components/tabs/ScatterPlotTab.tsx

```diff
@@ -0,0 +1,709 @@
+  }, [plotRef.current, onClickPlot]);
+
+  //////////////////////////////////
+  // Helper Methods
+  //////////////////////////////////
+
+  /** Retrieve feature data, if it exists. Accounts for the artificially-added time feature. */
+  const getData = useCallback(
+    (featureName: string | null, dataset: Dataset | null): Uint32Array | Float32Array | undefined => {
+      if (featureName === null || dataset === null) {
+        return undefined;
+      }
+      if (featureName === TIME_FEATURE.name) {
+        return dataset.times || undefined;
+      }
+      return dataset.getFeatureData(featureName)?.data;
+    },
+    []
+  );
+
+  /**
+   * Returns a hex color string with increasing transparency as the number of markers increases.
+   * Base color must be a 7-character RGB hex string (e.g. `#000000`).
+   */
+  const applyMarkerTransparency = (numMarkers: number, baseColor: string): string => {
+    if (baseColor.length !== 7) {
+      throw new Error("ScatterPlotTab.getMarkerColor: Base color '" + baseColor + "' must be 7-character hex string.");
+    }
+    // Interpolate linearly between 80% and 25% transparency from 0 up to a max of 1000 markers.
+    const opacity = remap(numMarkers, 0, 1000, 0.8, 0.25);
+    return (
+      baseColor +
+      Math.floor(opacity * 255)
+        .toString(16)
+        .padStart(2, "0")
+    );
+  };
+
+  // Track last rendered props + state to make optimizations on re-renders
+  type LastRenderedState = {
+    rangeType: RangeType;
+    xAxisFeatureName: string | null;
+    yAxisFeatureName: string | null;
+  } & ScatterPlotTabProps;
+
+  const lastRenderedState = useRef<LastRenderedState>({
+    rangeType: DEFAULT_RANGE_TYPE,
+    xAxisFeatureName: null,
+    yAxisFeatureName: null,
+    ...props,
+  });
+
+  /**
+   * Returns an object with flags indicating which props have changed since the last render.
+   */
+  const getChangeFlags = (): {
+    haveAxesChanged: boolean;
+    hasRangeChanged: boolean;
+    hasTrackChanged: boolean;
+    hasFrameChanged: boolean;
+    hasDatasetChanged: boolean;
+  } => {
+    const lastState = lastRenderedState.current;
+    return {
+      haveAxesChanged:
+        lastState.xAxisFeatureName !== xAxisFeatureName || lastState.yAxisFeatureName !== yAxisFeatureName,
+      hasRangeChanged: lastState.rangeType !== rangeType,
+      hasTrackChanged: lastState.selectedTrack !== selectedTrack,
+      hasFrameChanged: lastState.currentFrame !== currentFrame,
+      hasDatasetChanged: lastState.dataset !== dataset,
+    };
+  };
+
+  /** Returns whether the changes would result in a new plot, requiring the zoom and UI to reset. */
+  const shouldPlotUiReset = (): boolean => {
+    // Only reset the plot if the axes or range have changed, ignoring the track or frame.
+    // This prevents clicking on a new track from resetting the plot view.
+    const flags = getChangeFlags();
+    return flags.haveAxesChanged || flags.hasRangeChanged || flags.hasDatasetChanged;
+  };
```

Comment:
I originally was using more of these flags for optimization and left it in assuming I'd need it, but since adding coloring to the scatterplot (and the various performance benefits that has introduced), I don't think I will. Thanks for calling this one out, I've refactored this.